lib/metrics_target_methods.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+require "shellwords"
+
+module MetricsTargetMethods
+  MAX_SCRAPE_FETCH_COUNT = 4
+  FILENAME_FORMAT = "%Y-%m-%dT%H-%M-%S-%N"
+
+  def metrics_config
+    {
+      # Array of endpoints to collect metrics from
+      endpoints: [],
+
+      # Maximum number of entries in the pending buffer
+      max_pending_buffer_size: 120,
+
+      # Interval for collecting metrics in seconds or as a time span string
+      interval: "15s",
+
+      # Additional label names and values to be added to the collected metrics
+      additional_labels: {foo: "bar"},
+
+      # Directory to store the collected metrics
+      metrics_dir: "/home/ubi/metrics"
+    }
+  end
+
+  def export_metrics(session:, tsdb_client:)
+    scrape_results = scrape_endpoints(session)
+
+    if scrape_results.empty?
+      return
+    end
+
+    scrape_results.each do |scrape|
+      tsdb_client.import_prometheus(scrape, metrics_config[:additional_labels])
+    end
+
+    mark_pending_scrapes_as_done(session, scrape_results[-1].time)
+    scrape_results.count
+  end
+
+  def scrape_endpoints(session)
+    scrape_files = session[:ssh_session].exec!("ls #{metrics_dir}/pending | sort | head -n #{MAX_SCRAPE_FETCH_COUNT}").split("\n")
+
+    scrape_files.filter_map do |file|
+      time_str = file.split(".").first
+      time = Time.strptime(time_str, FILENAME_FORMAT)
+      status = {}
+
+      scrape_content = session[:ssh_session].exec!("cat #{metrics_dir}/pending/#{file.shellescape}", status: status)
+
+      VictoriaMetrics::Client::Scrape.new(time: time, samples: scrape_content) if status[:exit_code] == 0
```

Comment:
This class doesn't seem to add much value. You can use hash instead.

```
{time:, samples: scrape_content} if status[:exit_code] == 0
```

You can also omit repeating the second field by renaming it to `samples` in previous line.

Reply:
I changed this to use the `Data` core class